### ISSUE TITLE
Provide log parameters sanitisation as a service

### ISSUE THREAD

**Issue by kptdobe** [adobe/openwhisk-action-utils#27]
Consumers of the `openwhisk-action-utils` API have to worry about what they log and sanitise their secret. This will either end-up into code duplication or secrets being exposed in the logs because they will forget. 
For example: https://github.com/adobe/helix-dispatch/blob/master/src/index.js#L51-L58

This was supposed to be addressed by https://github.com/adobe/helix-log/issues/44 but since I managed to output many secrets yesterday, I think this is not properly supported and at least for sure it is not tested.

I do not know if this should be in https://github.com/adobe/helix-log or in this project, I leave to our "log architects" (@koraa @tripodsan @trieloff) to decide what is the best location for this. 

My only requirement is that when I log: 
```
{
  MY_SECRET_KEY: 'mykeyvalue'
} 
```

log outputs is:

```
{
  MY_SECRET_KEY: 'undisclosed secret'
} 
```

The text does not matter but the key should not be removed so that I know it is there but I cannot read the value! If we remove the key (like here https://github.com/adobe/openwhisk-action-utils/blob/master/src/logger.js#L205), I do not know if the key was not transported to the action, if I forgot it in the service config or if I use a format which is filtered out for a param to not filter like `MY_DEBUG_PARAM`.

**Comment by koraa** [comment]
JFYI: adobe/helix-log#44 provides a way to tag secrets so they are not logged…did you tag the secrets in such a manner?

**Comment by kptdobe** [comment]
You mean append a `.secret` to all secrets key ? This logic would need to be added somewhere (which might be the solution for this issue).

But as a consumer of `openwhisk-action-utils` I do not think I want that logging: 

```
{
  MY_SECRET_KEY: 'mykeyvalue'
}
```

outputs 

```
{
  MY_SECRET_KEY.secret: 'undisclosed secret'
}
````

If the .secret does not appear in the output, then this would be ok.
Then the task would be here to use the current regex (all uppercase keys) and append the `.secret` when logging.

But it might be easier to provide the regex to the logger and each key matching the regex would be considered as a secret.

**Comment by koraa** [comment]
Ah no, I mean the `Secret` wrapper class. https://github.com/adobe/helix-log/blob/master/src/secret.js

It is designed to prevent any usage of the value unless the dev explicitly states "please extract the secret so I can use it". By wrapping all secret values in such a class, no secret should be exposed.

**Comment by tripodsan** [comment]
yes..we could use the `Secret` value wrapper, but we still need a service that detects all the input params of the action as _secret_ and replaces them with the `Secret` value wrapper.

**Comment by tripodsan** [comment]
also see #25

**Comment by tripodsan** [comment]
IMO, we should move all _secrets_ to a nested action parameter, so we can hide them at once....

**Comment by koraa** [comment]
@tripodsan I don't think you need a replacer service. Secret is tested and proven to hide secrets on (1) JSON conversion (2) console.log (3) logging with helix-log and should work in most other circumstances (it works internally by creating a hidden/non-enumerable property as a getter/setter pair). Unless one is using Object.getProperties, no implicit access should leak the secret.

**Comment by tripodsan** [comment]
> Don't we know which properties are secret?

a good heuristic is:
1. they only have uppercase letters
2. they contain words like: `KEY`, `PASS`, `PWD`, `PASSWORD`, `PRIVATE`, `TOKEN`, `SECRET`

**Comment by trieloff** [comment]
3. they have values that are long strings of `/[0-9a-f-_]{16}/`

**Comment by koraa** [comment]
Why not just create a list of those properties that are secrets? This should be known for any action…

**Comment by trieloff** [comment]
We add the secrets at the package level, so adding a secret that is needed for one action will make it visible for all other actions in the same package. You don't want to change their code – or even have to be aware of their existence, so adhering to the convention that ALL_CAPS_STUFF is secret is a good idea.

**Comment by tripodsan** [comment]
> 3. they have values that are long strings of `/[0-9a-f-_]{16}/`

like an activation id ?? ;-)

**Comment by trieloff** [comment]
That's something for the whitelist…
